src/RamTransforms.cpp

```diff
@@ -295,4 +295,115 @@ bool IfConversionTransformer::convertIndexScans(RamProgram& program) {
     return changed;
 }
 
+std::unique_ptr<RamOperation> ChoiceConversionTransformer::rewriteScan(const RamScan* scan) {
+    bool transformTuple = false;
+
+    // Check that RamFilter follows the Scan in the loop nest
+    if (const auto* filter = dynamic_cast<const RamFilter*>(&scan->getOperation())) {
+        // Check that the Filter uses the identifier in the Scan
+        if (rcla->getLevel(&filter->getCondition()) == scan->getTupleId()) {
+            transformTuple = true;
+
+            // Check that the filter is not referred to after
+            const auto* nextNode = dynamic_cast<const RamNode*>(&filter->getOperation());
+            visitDepthFirst(*nextNode, [&](const RamNode& node) {
```

Comment:
This is an odd approach. Why not use two visitors, one for RamElementAccess and one for RamUnpackRecord?

Reply:
A traversal is not cheap - and it is cheaper to check the cases inside a single visitor.

Reply:
With two visitors the purpose of the code is clearer, and the use of visitDepthFirst is more standard, thus quicker to read.

That said, the complexity of two traversals that do one operation is the same as one traversal that does two operations, surely? In this case we're performing 50% more dynamic casts if we combine the two vs double the iteration costs if they are separate, so any improvement in constant factors seems arguable. The RAM is also not large, so performance is not an issue here.